### PR TITLE
ENT-5115 Added checks for database state to avoid error messages during install

### PR DESCRIPTION
**Pull request by craigcomstock**
Changelog: title
Ticket: ENT-5115

merge together:
https://github.com/cfengine/nova/pull/1573
https://github.com/cfengine/buildscripts/pull/659
https://github.com/cfengine/mission-portal/pull/988

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @Lex-2008 can review this?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 1]
```diff
@@ -716,66 +716,70 @@ then
...
+      if ! su cfpostgres -c "$PREFIX/bin/psql  -d postgres -c '\du' | grep cfmppostgres >/dev/null"; then
+        su cfpostgres -c "$PREFIX/bin/psql cfmp" < $PREFIX/share/GUI/phpcfenginenova/create_cfmppostgres_user.sql
+      fi
+  )
 
   # Ensure cfpostgres can read the sql files it will import. And that they are
   # restored to restrictive state after import ENT-2684
   chown cfpostgres $PREFIX/share/db/*.sql
-  (cd /tmp && chown cfpostgres "$PREFIX/share/db/schema.sql" && su cfpostgres -c "$PREFIX/bin/psql cfdb -v ON_ERROR_STOP=1 -f $PREFIX/share/db/schema.sql" && chown root "$PREFIX/share/db/schema.sql")
+  (cd /tmp && chown cfpostgres "$PREFIX/share/db/schema.sql" && su cfpostgres -c "$PREFIX/bin/psql cfdb -f $PREFIX/share/db/schema.sql" && chown root "$PREFIX/share/db/schema.sql" || false)
```
Comment (Lex-2008) [inline packaging/common/cfengine-hub/postinstall.sh]: is `|| false` needed here? It gets executed if previous command resulted in FAIL return code, and without it previous command return code will be outputted. Or am I missing something?

Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: I added it for a reason to be sure. I will double-check though.

Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: Agreed. I don't know why I did this. I will remove `|| false` and re-run through jenkins. As @vpodzime mentioned in-person the `|| false` likely just accomplishes a return of error code 1 always. :p

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4371/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4371/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4371/

File: packaging/common/cfengine-hub/postinstall.sh [hunk 2]
```diff
@@ -716,66 +716,70 @@ then
...
     cd /tmp
-    set +e
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfdb"
-    su cfpostgres -c "$PREFIX/bin/createuser -S -D -R -w $MP_APACHE_USER"
-    su cfpostgres -c "$PREFIX/bin/createuser -d -a -w root"
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfmp"
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfsettings"
-    exit 0
+    for db_name in cfdb cfmp cfsettings; do
+      if ! su cfpostgres -c "$PREFIX/bin/psql --list | grep ${db_name} >/dev/null"; then
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: Please get the list once and then search for DB names in it.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 3]
```diff
@@ -716,66 +716,70 @@ then
...
+        su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 ${db_name}"
+      fi
+    done
+    if ! su cfpostgres -c "$PREFIX/bin/psql  -d postgres -c '\du' | grep $MP_APACHE_USER >/dev/null"; then
+      su cfpostgres -c "$PREFIX/bin/createuser -S -D -R -w $MP_APACHE_USER"
+    fi
+    if ! su cfpostgres -c "$PREFIX/bin/psql  -d postgres -c '\du' | grep root >/dev/null"; then
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: Same here with the list of users.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 4]
```diff
@@ -716,66 +716,70 @@ then
...
-    chown cfpostgres "$PREFIX/share/db/ootb_settings.sql" && su cfpostgres -c "$PREFIX/bin/psql cfsettings  -v ON_ERROR_STOP=1 -f $PREFIX/share/db/ootb_settings.sql" && chown root "$PREFIX/share/db/ootb_settings.sql"
-    chown cfpostgres "$PREFIX/share/db/ootb_import.sql" && su cfpostgres -c "$PREFIX/bin/psql cfdb  -v ON_ERROR_STOP=1 -f $PREFIX/share/db/ootb_import.sql" && chown root "$PREFIX/share/db/ootb_import.sql"
+    chown cfpostgres "$PREFIX/share/db/schema_settings.sql" && su cfpostgres -c "$PREFIX/bin/psql cfsettings -f $PREFIX/share/db/schema_settings.sql" && chown root "$PREFIX/share/db/schema_settings.sql" || false
+    chown cfpostgres "$PREFIX/share/db/ootb_settings.sql" && su cfpostgres -c "$PREFIX/bin/psql cfsettings -f $PREFIX/share/db/ootb_settings.sql" && chown root "$PREFIX/share/db/ootb_settings.sql" || false
+    chown cfpostgres "$PREFIX/share/db/ootb_import.sql" && su cfpostgres -c "$PREFIX/bin/psql cfdb -f $PREFIX/share/db/ootb_import.sql" && chown root "$PREFIX/share/db/ootb_import.sql" || false
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: This reverts a recent change which added the `ON_ERROR_STOP`, right? Is it really necessary?

Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: Right. I'll make sure ON_ERROR_STOP=1 is present.

Comment (craigcomstock) [reply packaging/common/cfengine-hub/postinstall.sh]: Yes. ON_ERROR_STOP=1 is in place in each individual sql file now so not needed here.

**Comment by craigcomstock** [comment]
Things look good. I tested 3.12.2 -> PR as well as fresh install with PR.
Tomorrow I will test some more failure cases (like cfdb missing, really odd case) as well as 3.10.latest -> PR.

**Comment by craigcomstock** [comment]
@cf-bottom jenkins again please, for me to test tomorrow. :+1:

**Comment by craigcomstock** [comment]
@vpodzime take another look, have addressed your concerns I think. Will test a bit more tomorrow with PR builds before merging.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4412/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4412/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4412/

File: packaging/common/cfengine-hub/postinstall.sh [hunk 5]
```diff
@@ -718,66 +718,76 @@ then
...
     cd /tmp
-    set +e
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfdb"
-    su cfpostgres -c "$PREFIX/bin/createuser -S -D -R -w $MP_APACHE_USER"
-    su cfpostgres -c "$PREFIX/bin/createuser -d -a -w root"
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfmp"
-    su cfpostgres -c "$PREFIX/bin/createdb -E SQL_ASCII --lc-collate=C --lc-ctype=C -T template0 cfsettings"
-    exit 0
+
+    su cfpostgres -c "$PREFIX/bin/psql --list" > db_name_list.txt
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: I'm not sure writing all this information into files and especially under `/tmp` where every user can read those files is a good idea. You can use `db_name_list=$(su cfpostgres...); if echo "$db_name_list" | grep ...; then` instead.

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4498/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4498/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4498/

**Pull request merged**